optimizer/fold.go

```diff
@@ -211,6 +215,50 @@ func (fold *fold) Visit(node *Node) {
 					patchWithType(&FloatNode{Value: math.Pow(a.Value, b.Value)}, a.Type())
 				}
 			}
+		case "and", "&&":
+			a := toBool(n.Left)
+			b := toBool(n.Right)
+
+			if a != nil && a.Value { // true and x
+				patch(n.Right)
+			} else if b != nil && b.Value { // x and true
+				patch(n.Left)
+			} else if (a != nil && !a.Value) || (b != nil && !b.Value) { // "x and false" or "false and x"
+				patch(&BoolNode{Value: false})
+			}
+		case "or", "||":
+			a := toBool(n.Left)
+			b := toBool(n.Right)
+
+			if a != nil && !a.Value { // false or x
+				patch(n.Right)
+			} else if b != nil && !b.Value { // x or false
+				patch(n.Left)
+			} else if (a != nil && a.Value) || (b != nil && b.Value) { // "x or true" or "true or x"
+				patch(&BoolNode{Value: true})
+			}
+		case "==":
+			{
+				a := toInteger(n.Left)
+				b := toInteger(n.Right)
+				if a != nil && b != nil {
+					patch(&BoolNode{Value: a.Value == b.Value})
+				}
+			}
```

Comment:
I skipped adding float/float and int/float precomputation here given the potential issues with floating point comparison, but I could add it in if desired.